web-module/src/main/webapp/resources/structEditJS/autocomplete.mjs

```diff
@@ -139,35 +139,95 @@ class CelAutocompleteInitialiser {
     selectElem.innerHTML = '<option selected="selected" value="">delete</option>';
   }
   
+  #handleAddNewMessage(event, selectElem) {
+    console.debug('addNewButton popup returned', event.data);
+    event.source?.close();
+    const option = new Option(event.data.text, event.data.id, true, true);
+    selectElem.append(option);
+    selectElem.dispatchEvent(new Event("change", {
+      view: window,
+      bubbles: true,
+      cancelable: true,
+    }));
+  }
+
+  #addNewButtonClickHandler(event, selectElem) {
+    const buttonElem = event.target;
+    const addNewUrl = buttonElem.getAttribute('data-url');
+    console.debug('addNewButtonClickHandler start', event, selectElem, addNewUrl);
+    const theAddNewPopup = window.open(addNewUrl, '_blank', 'popup=true');
+    theAddNewPopup.addEventListener('message', (ev) => this.#handleAddNewMessage(ev, selectElem));
+  }
+
+  #addNewButtonElem(selectElem, addNewUrl) {
+    const classField = selectElem.dataset.classField || ''; 
+    const type = selectElem.dataset.autocompleteType || '';
+    const cellRef = selectElem.dataset.cellRef || '';
+    const buttonText = window.celMessages.structEditor.autocomplete['addNewButtonText_' + cellRef]
+                || window.celMessages.structEditor.autocomplete['addNewButtonText_' + classField]
+                || window.celMessages.structEditor.autocomplete['addNewButtonText_' + type]
+                || window.celMessages.structEditor.autocomplete['addNewButtonText_default']
+                || 'nothing found? add new';
+    const iconElem = document.createElement('span');
+    iconElem.classList.add('halflings', 'halflings-plus-sign');
+    const buttonInnerElem = document.createElement('div');
+    buttonInnerElem.appendChild(iconElem);
+    buttonInnerElem.classList.add('view_cel_buttonLink');
+    buttonInnerElem.setAttribute('data-url', addNewUrl);
+    buttonInnerElem.insertAdjacentText('beforeend', buttonText)
+    buttonInnerElem.addEventListener('click', 
+        (ev) => this.#addNewButtonClickHandler(ev, selectElem));
+    const buttonElem = document.createElement('div');
+    buttonElem.appendChild(buttonInnerElem);
+    buttonElem.classList.add('box', 'cel_button', 'struct_autocomplete_addnew');
+    const itemElem = document.createElement('div');
+    itemElem.appendChild(buttonElem);
+    itemElem.classList.add('result', 'clearfix');
+    return itemElem;
+  }
+
+  #concatAddNewButtonToResults(selectElem, response) {
+    const results = response.results || [];
+    const addNewUrl = response.addNewUrl || '';
+    if (!response.hasMore && (addNewUrl !== '')) {
```

Comment:
shouldn't the button always show if it is configured even if there are more results?

Reply:
The button is showing as soon as the user has validated that it can not be found. Therefore the button shows up as last element in the result list.